// openbim-components/src/civil/CivilNavigator/src/curve-highlighter.js
import * as THREE from "three";
import { Line2 } from "three/examples/jsm/lines/Line2";
import { LineGeometry } from "three/examples/jsm/lines/LineGeometry";
import { LineMaterial } from "three/examples/jsm/lines/LineMaterial";
import { Event } from "../../../base-types";
export class CurveHighlighter {
    constructor(scene, type) {
        this.onSelect = new Event();
        this.caster = new THREE.Raycaster();
        this.scene = scene;
        this.type = type;
        this.hoverCurve = this.newCurve(0.003, 0x444444, false);
        this.hoverPoints = this.newPoints(5, 0x444444);
        this.selectCurve = this.newCurve(0.005, 0xffffff, true);
        this.selectPoints = this.newPoints(7, 0xffffff);
    }
    dispose() {
        if (this.selectCurve) {
            this.scene.remove(this.selectCurve);
        }
        this.selectCurve.material.dispose();
        this.selectCurve.geometry.dispose();
        this.selectCurve = null;
        this.hoverCurve.material.dispose();
        this.hoverCurve.geometry.dispose();
        this.hoverCurve = null;
        this.hoverPoints.material.dispose();
        this.hoverPoints.geometry.dispose();
        this.selectPoints.material.dispose();
        this.selectPoints.geometry.dispose();
        this.scene = null;
    }
    castRay(event, camera, dom, meshes) {
        const mouse = new THREE.Vector2();
        const rect = dom.getBoundingClientRect();
        mouse.x = ((event.clientX - rect.left) / rect.width) * 2 - 1;
        mouse.y = -((event.clientY - rect.top) / rect.height) * 2 + 1;
        this.caster.setFromCamera(mouse, camera);
        const intersects = this.caster.intersectObjects(meshes);
        if (!intersects.length) {
            return null;
        }
        return intersects[0];
    }
    select(mesh) {
        this.highlight(mesh, this.selectCurve, this.selectPoints, true);
        this.onSelect.trigger(mesh);
    }
    unSelect() {
        this.selectCurve.removeFromParent();
        this.selectPoints.removeFromParent();
    }
    hover(mesh) {
        this.highlight(mesh, this.hoverCurve, this.hoverPoints, false);
    }
    unHover() {
        this.hoverCurve.removeFromParent();
        this.hoverPoints.removeFromParent();
    }
    highlight(mesh, curve, points, useColors) {
        const { alignment } = mesh.curve;
        this.scene.add(curve);
        this.scene.add(points);
        const lines = [];
        const colors = [];
        const vertices = [];
        for (const foundCurve of alignment[this.type]) {
            const position = foundCurve.mesh.geometry.attributes.position;
            for (const coord of position.array) {
                lines.push(coord);
            }
            if (useColors) {
                let type;
                if (this.type === "absolute") {
                    // 3D curves don't have type defined, so we take the horizontal
                    const { horizontal } = foundCurve.alignment;
                    type = horizontal[foundCurve.index].data.TYPE;
                }
                else {
                    type = foundCurve.data.TYPE;
                }
                const found = CurveHighlighter.settings.colors[type] || [1, 1, 1];
                for (let i = 0; i < position.count; i++) {
                    colors.push(...found);
                }
            }
            const [x, y, z] = position.array;
            vertices.push(new THREE.Vector3(x, y, z));
        }
        const lastX = lines[lines.length - 3];
        const lastY = lines[lines.length - 2];
        const lastZ = lines[lines.length - 1];
        vertices.push(new THREE.Vector3(lastX, lastY, lastZ));
        if (lines.length / 3 > curve.geometry.attributes.position.count) {
            curve.geometry.dispose();
            curve.geometry = new LineGeometry();
        }
        curve.geometry.setPositions(lines);
        if (useColors) {
            curve.geometry.setColors(colors);
        }
        points.geometry.setFromPoints(vertices);
    }
    newCurve(linewidth, color, vertexColors) {
        const selectGeometry = new LineGeometry();
        const selectMaterial = new LineMaterial({
            color,
            linewidth,
            vertexColors,
            worldUnits: false,
            depthTest: false,
        });
        const curve = new Line2(selectGeometry, selectMaterial);
        this.scene.add(curve);
        return curve;
    }
    newPoints(size, color) {
        const pointsGeometry = new THREE.BufferGeometry();
        const pointsAttr = new THREE.BufferAttribute(new Float32Array(), 3);
        pointsGeometry.setAttribute("position", pointsAttr);
        const pointsMaterial = new THREE.PointsMaterial({
            size,
            color,
            sizeAttenuation: false,
            depthTest: false,
        });
        const points = new THREE.Points(pointsGeometry, pointsMaterial);
        points.frustumCulled = false;
        this.scene.add(points);
        return points;
    }
}
CurveHighlighter.settings = {
    colors: {
        LINE: [213 / 255, 0 / 255, 255 / 255],
        CIRCULARARC: [0 / 255, 46, 255 / 255],
        CLOTHOID: [0 / 255, 255 / 255, 0 / 255],
        PARABOLICARC: [0 / 255, 255 / 255, 72 / 255],
        CONSTANTGRADIENT: [213 / 255, 0 / 255, 255 / 255],
    },
};
