// openbim-components/src/civil/CivilNavigator/index.js
import * as THREE from "three";
import { Component, Event } from "../../base-types";
import { Simple2DMarker, Simple2DScene } from "../../core";
export class CivilNavigator extends Component {
    constructor(components) {
        super(components);
        this.enabled = true;
        this.onHighlight = new Event();
        this.onMarkerChange = new Event();
        this.onMarkerHidden = new Event();
        this._curveMeshes = [];
        this._previousAlignment = null;
        this.scene = new Simple2DScene(this.components, false);
        this.mouseMarkers = {
            select: this.newMouseMarker("#ffffff"),
            hover: this.newMouseMarker("#575757"),
        };
        this.setupEvents();
        this.adjustRaycasterOnZoom();
    }
    initialize() {
        console.log("View for CivilNavigator: ", this.view);
    }
    get() {
        return null;
    }
    async draw(model, filter) {
        if (!model.civilData) {
            throw new Error("The provided model doesn't have civil data!");
        }
        const { alignments } = model.civilData;
        const allAlignments = filter || alignments.values();
        const scene = this.scene.get();
        const totalBBox = new THREE.Box3();
        totalBBox.makeEmpty();
        totalBBox.min.set(Number.MAX_VALUE, Number.MAX_VALUE, Number.MAX_VALUE);
        totalBBox.max.set(-Number.MAX_VALUE, -Number.MAX_VALUE, -Number.MAX_VALUE);
        for (const alignment of allAlignments) {
            if (!alignment) {
                throw new Error("Alignment not found!");
            }
            for (const curve of alignment[this.view]) {
                scene.add(curve.mesh);
                this._curveMeshes.push(curve.mesh);
                if (!totalBBox.isEmpty()) {
                    totalBBox.expandByObject(curve.mesh);
                }
                else {
                    curve.mesh.geometry.computeBoundingBox();
                    const cbox = curve.mesh.geometry.boundingBox;
                    if (cbox instanceof THREE.Box3) {
                        totalBBox.copy(cbox).applyMatrix4(curve.mesh.matrixWorld);
                    }
                }
            }
        }
        const scaledBbox = new THREE.Box3();
        const size = new THREE.Vector3();
        const center = new THREE.Vector3();
        totalBBox.getCenter(center);
        totalBBox.getSize(size);
        size.multiplyScalar(1.2);
        scaledBbox.setFromCenterAndSize(center, size);
        await this.scene.controls.fitToBox(scaledBbox, false);
    }
    setupEvents() {
        this.scene.uiElement
            .get("container")
            .domElement.addEventListener("mousemove", async (event) => {
            const dom = this.scene.uiElement.get("container").domElement;
            const result = this.highlighter.castRay(event, this.scene.camera, dom, this._curveMeshes);
            if (result) {
                const { object } = result;
                this.highlighter.hover(object);
                await this.updateMarker(result, "hover");
                return;
            }
            this.mouseMarkers.hover.visible = false;
            this.highlighter.unHover();
            await this.onMarkerHidden.trigger({ type: "hover" });
        });
        this.scene.uiElement
            .get("container")
            .domElement.addEventListener("click", async (event) => {
            const dom = this.scene.uiElement.get("container").domElement;
            const intersects = this.highlighter.castRay(event, this.scene.camera, dom, this._curveMeshes);
            if (intersects) {
                const result = intersects;
                const mesh = result.object;
                this.highlighter.select(mesh);
                await this.updateMarker(result, "select");
                await this.onHighlight.trigger({ mesh, point: result.point });
                if (this._previousAlignment !== mesh.curve.alignment) {
                    this.kpManager.clearKPStations();
                    // this.showKPStations(mesh);
                    this.kpManager.showKPStations(mesh);
                    // this.kpManager.createKP();
                    this._previousAlignment = mesh.curve.alignment;
                }
            }
            // this.highlighter.unSelect();
            // this.clearKPStations();
        });
    }
    async dispose() {
        this.highlighter.dispose();
        this.clear();
        this.onHighlight.reset();
        await this.scene.dispose();
        this._curveMeshes = [];
    }
    clear() {
        this.highlighter.unSelect();
        this.highlighter.unHover();
        for (const mesh of this._curveMeshes) {
            mesh.removeFromParent();
        }
        this._curveMeshes = [];
    }
    setMarker(alignment, percentage, type) {
        if (!this._curveMeshes.length) {
            return;
        }
        const found = alignment.getCurveAt(percentage, this.view);
        const point = alignment.getPointAt(percentage, this.view);
        const { index } = found.curve.getSegmentAt(found.percentage);
        this.setMouseMarker(point, found.curve.mesh, index, type);
    }
    setDefSegments(segmentsArray) {
        const defSegments = [];
        const slope = [];
        const calculateSlopeSegment = (point1, point2) => {
            const deltaY = point2[1] - point1[1];
            const deltaX = point2[0] - point1[0];
            return deltaY / deltaX;
        };
        for (let i = 0; i < segmentsArray.length; i++) {
            const segment = segmentsArray[i];
            let startX;
            let startY;
            let endX;
            let endY;
            // Set start
            for (let j = 0; j < Object.keys(segment).length / 3; j++) {
                if (segment[j * 3] !== undefined && segment[j * 3 + 1] !== undefined) {
                    startX = segment[j * 3];
                    startY = segment[j * 3 + 1];
                    break;
                }
            }
            // Set end
            for (let j = Object.keys(segment).length / 3 - 1; j >= 0; j--) {
                if (segment[j * 3] !== undefined && segment[j * 3 + 1] !== undefined) {
                    endX = segment[j * 3];
                    endY = segment[j * 3 + 1];
                    break;
                }
            }
            const defSlope = calculateSlopeSegment(
            // @ts-ignore
            [startX, startY], 
            // @ts-ignore
            [endX, endY]);
            const slopeSegment = (defSlope * 100).toFixed(2);
            slope.push({ slope: slopeSegment });
        }
        for (const segment of segmentsArray) {
            for (let i = 0; i < segment.length - 3; i += 3) {
                const startX = segment[i];
                const startY = segment[i + 1];
                const startZ = segment[i + 2];
                const endX = segment[i + 3];
                const endY = segment[i + 4];
                const endZ = segment[i + 5];
                defSegments.push({
                    start: new THREE.Vector3(startX, startY, startZ),
                    end: new THREE.Vector3(endX, endY, endZ),
                });
            }
        }
        return { defSegments, slope };
    }
    hideMarker(type) {
        this.mouseMarkers[type].visible = false;
    }
    adjustRaycasterOnZoom() {
        this.scene.controls.addEventListener("update", () => {
            const { zoom, left, right, top, bottom } = this.scene.camera;
            const width = left - right;
            const height = top - bottom;
            const screenSize = Math.max(width, height);
            const realScreenSize = screenSize / zoom;
            const range = 40;
            const { caster } = this.highlighter;
            caster.params.Line.threshold = realScreenSize / range;
        });
    }
    newMouseMarker(color) {
        const scene = this.scene.get();
        const root = document.createElement("div");
        const bar = document.createElement("div");
        root.appendChild(bar);
        bar.style.backgroundColor = color;
        bar.style.width = "3rem";
        bar.style.height = "3px";
        const mouseMarker = new Simple2DMarker(this.components, root, scene);
        mouseMarker.visible = false;
        return mouseMarker;
    }
    setMouseMarker(point, object, index, type) {
        if (index === undefined) {
            return;
        }
        this.mouseMarkers[type].visible = true;
        const marker = this.mouseMarkers[type].get();
        marker.position.copy(point);
        const curveMesh = object;
        const { startPoint, endPoint } = curveMesh.curve.getSegment(index);
        const angle = Math.atan2(endPoint.y - startPoint.y, endPoint.x - startPoint.x);
        const bar = marker.element.children[0];
        const trueAngle = 90 - (angle / Math.PI) * 180;
        bar.style.transform = `rotate(${trueAngle}deg)`;
    }
    async updateMarker(intersects, type) {
        const { point, index, object } = intersects;
        const mesh = object;
        const curve = mesh.curve;
        const alignment = mesh.curve.alignment;
        const percentage = alignment.getPercentageAt(point, this.view);
        const markerPoint = point.clone();
        this.setMouseMarker(markerPoint, mesh, index, type);
        if (percentage !== null) {
            await this.onMarkerChange.trigger({ alignment, percentage, type, curve });
        }
    }
}
